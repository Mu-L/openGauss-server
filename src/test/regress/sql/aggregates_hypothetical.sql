create schema aggregate;
set current_schema='aggregate';

create table test_aggregate(c1 int, c2 NUMBER(8,2), c3 varchar(20), c4 timestamp);

insert into test_aggregate values(1,0.1,'1','2024-09-01 09:22:00'),
(2,0.2,'2','2024-09-02 09:22:00'),
(3,0.1,'3','2024-09-03 09:22:00'),
(3,0.2,'3','2024-09-04 09:22:00'),
(3,0.3,'3','2024-09-05 09:22:00'),
(3,0.3,'3','2024-09-05 09:22:00'),
(4,0.2,'4','2024-09-06 09:22:00'),
(5,0.2,'5','2024-09-07 09:22:00'),
(6,0.2,'6','2024-09-08 09:22:00'),
(7,0.2,'7','2024-09-09 09:22:00'),
(8,0.2,'8','2024-09-10 09:22:00'),
(9,0.2,'9','2024-09-11 09:22:00'),
(10,0.2,'10','2024-09-12 09:22:00');

--cume_dist 
select cume_dist(3,0.2) within group (order by c1,c2) from test_aggregate;

--percent_rank
select percent_rank(3,0.2) within group (order by c1,c2) from test_aggregate;

--dense_rank
select dense_rank(4,0.2) within group (order by c1,c2) from test_aggregate;

--rank
select rank(4,0.2) within group (order by c1,c2) from test_aggregate;

-- divide by zero check
select percent_rank(0) within group (order by x) from generate_series(1,0) x;

--error, The number of parameters does not match
select cume_dist(3,0.2) within group (order by c1) from test_aggregate;
select cume_dist(3) within group (order by c1,c2) from test_aggregate;

-- error, ordered-set aggs can't use ungrouped vars in direct args:
select rank(x) within group (order by x) from generate_series(1,5) x;


-- enable_aggr_coerce_type = off,  type conversion test

--error
select cume_dist(3) within group (order by c3) from test_aggregate;

--error 
select cume_dist('a') within group (order by c1) from test_aggregate;

--error 
select cume_dist('2024') within group (order by c4) from test_aggregate;

--success
select cume_dist('2024-12-12') within group (order by c4) from test_aggregate;

--success
select cume_dist('1') within group (order by c1) from test_aggregate;

--error boolean
select rank(1) within group (order by x) from (values (true),(false)) v(x);

-- enable_aggr_coerce_type = on,  type conversion test

set enable_aggr_coerce_type = on;
--success
select cume_dist(3) within group (order by c3) from test_aggregate;

--error
select cume_dist('a') within group (order by c1) from test_aggregate;

--error
select cume_dist('2024') within group (order by c4) from test_aggregate;

--success
select cume_dist('2024-12-12') within group (order by c4) from test_aggregate;

--success
select cume_dist('1') within group (order by c1) from test_aggregate;

--sucdess boolean
select rank(1) within group (order by x) from (values (true),(false)) v(x);


set enable_aggr_coerce_type = off;
drop table test_aggregate;
drop schema aggregate CASCADE;
