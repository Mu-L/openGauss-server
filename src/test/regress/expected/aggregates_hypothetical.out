create schema aggregate;
set current_schema='aggregate';
create table test_aggregate(c1 int, c2 NUMBER(8,2), c3 varchar(20), c4 timestamp);
insert into test_aggregate values(1,0.1,'1','2024-09-01 09:22:00'),
(2,0.2,'2','2024-09-02 09:22:00'),
(3,0.1,'3','2024-09-03 09:22:00'),
(3,0.2,'3','2024-09-04 09:22:00'),
(3,0.3,'3','2024-09-05 09:22:00'),
(3,0.3,'3','2024-09-05 09:22:00'),
(4,0.2,'4','2024-09-06 09:22:00'),
(5,0.2,'5','2024-09-07 09:22:00'),
(6,0.2,'6','2024-09-08 09:22:00'),
(7,0.2,'7','2024-09-09 09:22:00'),
(8,0.2,'8','2024-09-10 09:22:00'),
(9,0.2,'9','2024-09-11 09:22:00'),
(10,0.2,'10','2024-09-12 09:22:00');
--cume_dist 
select cume_dist(3,0.2) within group (order by c1,c2) from test_aggregate;
    cume_dist     
------------------
 .357142857142857
(1 row)

select cume_dist(3,.05) within group (order by c1,c3) from test_aggregate;
    cume_dist     
------------------
 .214285714285714
(1 row)

--percent_rank
select percent_rank(3,0.2) within group (order by c1,c2) from test_aggregate;
   percent_rank   
------------------
 .230769230769231
(1 row)

select percent_rank(3,.05) within group (order by c1,c3) from test_aggregate;
   percent_rank   
------------------
 .153846153846154
(1 row)

--dense_rank
select dense_rank(4,0.2) within group (order by c1,c2) from test_aggregate;
 dense_rank 
------------
          6
(1 row)

select dense_rank(3,.05) within group (order by c1,c3) from test_aggregate;
 dense_rank 
------------
          3
(1 row)

--rank
select rank(4,0.2) within group (order by c1,c2) from test_aggregate;
 rank 
------
    7
(1 row)

select rank(3,.05) within group (order by c1,c3) from test_aggregate;
 rank 
------
    3
(1 row)

-- divide by zero check
select percent_rank(0) within group (order by x) from generate_series(1,0) x;
 percent_rank 
--------------
            0
(1 row)

--error, The number of parameters does not match
select cume_dist(3,0.2) within group (order by c1) from test_aggregate;
ERROR:  function cume_dist(integer, numeric, integer) does not exist
LINE 1: select cume_dist(3,0.2) within group (order by c1) from test...
               ^
HINT:  To use the hypothetical-set aggregate cume_dist, the number of hypothetical direct arguments (here 2) must match the number of ordering columns (here 1).
CONTEXT:  referenced column: cume_dist
select cume_dist(3) within group (order by c1,c2) from test_aggregate;
ERROR:  function cume_dist(integer, integer, numeric) does not exist
LINE 1: select cume_dist(3) within group (order by c1,c2) from test_...
               ^
HINT:  To use the hypothetical-set aggregate cume_dist, the number of hypothetical direct arguments (here 1) must match the number of ordering columns (here 2).
CONTEXT:  referenced column: cume_dist
-- error, ordered-set aggs can't use ungrouped vars in direct args:
select rank(x) within group (order by x) from generate_series(1,5) x;
ERROR:  column "x.x" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: select rank(x) within group (order by x) from generate_serie...
                    ^
DETAIL:  Direct arguments of an ordered-set aggregate must use only grouped columns.
-- enable_aggr_coerce_type = off,  type conversion test
select cume_dist(3) within group (order by c3) from test_aggregate;
    cume_dist     
------------------
 .571428571428571
(1 row)

--error 
select cume_dist('a') within group (order by c1) from test_aggregate;
ERROR:  invalid input syntax for integer: "a"
LINE 1: select cume_dist('a') within group (order by c1) from test_a...
                         ^
CONTEXT:  referenced column: cume_dist
--error 
select cume_dist('2024') within group (order by c4) from test_aggregate;
ERROR:  invalid input syntax for type timestamp: "2024"
LINE 1: select cume_dist('2024') within group (order by c4) from tes...
                         ^
CONTEXT:  referenced column: cume_dist
--success
select cume_dist('2024-12-12') within group (order by c4) from test_aggregate;
 cume_dist 
-----------
         1
(1 row)

--success
select cume_dist('1') within group (order by c1) from test_aggregate;
    cume_dist     
------------------
 .142857142857143
(1 row)

select rank(1) within group (order by x) from (values (true),(false)) v(x);
 rank 
------
    2
(1 row)

-- enable_aggr_coerce_type = on,  type conversion test
set enable_aggr_coerce_type = on;
--success
select cume_dist(3) within group (order by c3) from test_aggregate;
    cume_dist     
------------------
 .571428571428571
(1 row)

--error
select cume_dist('a') within group (order by c1) from test_aggregate;
ERROR:  invalid input syntax for integer: "a"
LINE 1: select cume_dist('a') within group (order by c1) from test_a...
                         ^
CONTEXT:  referenced column: cume_dist
--error
select cume_dist('2024') within group (order by c4) from test_aggregate;
ERROR:  invalid input syntax for type timestamp: "2024"
LINE 1: select cume_dist('2024') within group (order by c4) from tes...
                         ^
CONTEXT:  referenced column: cume_dist
--success
select cume_dist('2024-12-12') within group (order by c4) from test_aggregate;
 cume_dist 
-----------
         1
(1 row)

--success
select cume_dist('1') within group (order by c1) from test_aggregate;
    cume_dist     
------------------
 .142857142857143
(1 row)

--sucdess boolean
select rank(1) within group (order by x) from (values (true),(false)) v(x);
 rank 
------
    2
(1 row)

set enable_aggr_coerce_type = off;
drop table test_aggregate;
create table employees
(empno number,salary number,commission_pct varchar2(20));
SELECT RANK(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct desc) FROM employees;
 rank 
------
    1
(1 row)

SELECT RANK(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct desc nulls first) FROM employees;
 rank 
------
    1
(1 row)

SELECT RANK(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct desc nulls last) FROM employees;
 rank 
------
    1
(1 row)

SELECT RANK(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct asc) FROM employees;
 rank 
------
    1
(1 row)

SELECT RANK(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct asc nulls first) FROM employees;
 rank 
------
    1
(1 row)

SELECT RANK(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct asc nulls last) FROM employees;
 rank 
------
    1
(1 row)

SELECT RANK(15500, .05) WITHIN GROUP (ORDER BY salary desc, commission_pct) FROM employees;
 rank 
------
    1
(1 row)

SELECT RANK(15500, .05) WITHIN GROUP (ORDER BY salary desc nulls first, commission_pct) FROM employees;
 rank 
------
    1
(1 row)

SELECT RANK(15500, .05) WITHIN GROUP (ORDER BY salary desc nulls last, commission_pct) FROM employees;
 rank 
------
    1
(1 row)

SELECT RANK(15500, .05) WITHIN GROUP (ORDER BY salary asc, commission_pct) FROM employees;
 rank 
------
    1
(1 row)

SELECT RANK(15500, .05) WITHIN GROUP (ORDER BY salary asc nulls first, commission_pct) FROM employees;
 rank 
------
    1
(1 row)

SELECT RANK(15500, .05) WITHIN GROUP (ORDER BY salary asc nulls last, commission_pct) FROM employees;
 rank 
------
    1
(1 row)

SELECT dense_rank(15500, .05) WITHIN GROUP (ORDER BY salary desc nulls first, commission_pct) FROM employees;
 dense_rank 
------------
          1
(1 row)

SELECT dense_rank(15500, .05) WITHIN GROUP (ORDER BY salary desc nulls last, commission_pct) FROM employees;
 dense_rank 
------------
          1
(1 row)

SELECT dense_rank(15500, .05) WITHIN GROUP (ORDER BY salary asc, commission_pct) FROM employees;
 dense_rank 
------------
          1
(1 row)

SELECT dense_rank(15500, .05) WITHIN GROUP (ORDER BY salary asc nulls first, commission_pct) FROM employees;
 dense_rank 
------------
          1
(1 row)

SELECT dense_rank(15500, .05) WITHIN GROUP (ORDER BY salary asc nulls last, commission_pct) FROM employees;
 dense_rank 
------------
          1
(1 row)

SELECT PERCENT_RANK(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct desc) FROM employees;
 percent_rank 
--------------
            0
(1 row)

SELECT PERCENT_RANK(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct desc nulls first) FROM employees;
 percent_rank 
--------------
            0
(1 row)

SELECT PERCENT_RANK(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct desc nulls last) FROM employees;
 percent_rank 
--------------
            0
(1 row)

SELECT PERCENT_RANK(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct asc) FROM employees;
 percent_rank 
--------------
            0
(1 row)

SELECT PERCENT_RANK(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct asc nulls first) FROM employees;
 percent_rank 
--------------
            0
(1 row)

SELECT PERCENT_RANK(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct asc nulls last) FROM employees;
 percent_rank 
--------------
            0
(1 row)

SELECT PERCENT_RANK(15500, .05) WITHIN GROUP (ORDER BY salary desc, commission_pct) FROM employees;
 percent_rank 
--------------
            0
(1 row)

SELECT PERCENT_RANK(15500, .05) WITHIN GROUP (ORDER BY salary desc nulls first, commission_pct) FROM employees;
 percent_rank 
--------------
            0
(1 row)

SELECT PERCENT_RANK(15500, .05) WITHIN GROUP (ORDER BY salary desc nulls last, commission_pct) FROM employees;
 percent_rank 
--------------
            0
(1 row)

SELECT PERCENT_RANK(15500, .05) WITHIN GROUP (ORDER BY salary asc, commission_pct) FROM employees;
 percent_rank 
--------------
            0
(1 row)

SELECT PERCENT_RANK(15500, .05) WITHIN GROUP (ORDER BY salary asc nulls first, commission_pct) FROM employees;
 percent_rank 
--------------
            0
(1 row)

SELECT PERCENT_RANK(15500, .05) WITHIN GROUP (ORDER BY salary asc nulls last, commission_pct) FROM employees;
 percent_rank 
--------------
            0
(1 row)

SELECT CUME_DIST(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct desc) FROM employees;
 cume_dist 
-----------
         1
(1 row)

SELECT CUME_DIST(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct desc nulls first) FROM employees;
 cume_dist 
-----------
         1
(1 row)

SELECT CUME_DIST(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct desc nulls last) FROM employees;
 cume_dist 
-----------
         1
(1 row)

SELECT CUME_DIST(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct asc) FROM employees;
 cume_dist 
-----------
         1
(1 row)

SELECT CUME_DIST(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct asc nulls first) FROM employees;
 cume_dist 
-----------
         1
(1 row)

SELECT CUME_DIST(15500, .05) WITHIN GROUP (ORDER BY salary, commission_pct asc nulls last) FROM employees;
 cume_dist 
-----------
         1
(1 row)

SELECT CUME_DIST(15500, .05) WITHIN GROUP (ORDER BY salary desc, commission_pct) FROM employees;
 cume_dist 
-----------
         1
(1 row)

SELECT CUME_DIST(15500, .05) WITHIN GROUP (ORDER BY salary desc nulls first, commission_pct) FROM employees;
 cume_dist 
-----------
         1
(1 row)

SELECT CUME_DIST(15500, .05) WITHIN GROUP (ORDER BY salary desc nulls last, commission_pct) FROM employees;
 cume_dist 
-----------
         1
(1 row)

SELECT CUME_DIST(15500, .05) WITHIN GROUP (ORDER BY salary asc, commission_pct) FROM employees;
 cume_dist 
-----------
         1
(1 row)

SELECT CUME_DIST(15500, .05) WITHIN GROUP (ORDER BY salary asc nulls first, commission_pct) FROM employees;
 cume_dist 
-----------
         1
(1 row)

SELECT CUME_DIST(15500, .05) WITHIN GROUP (ORDER BY salary asc nulls last, commission_pct) FROM employees;
 cume_dist 
-----------
         1
(1 row)

drop table employees;
drop schema aggregate CASCADE;
